disk/disk.lisp

```diff
@@ -31,56 +28,101 @@
+  (flet ((read-cluster (disk addr n-sectors wired-buffer buffer base-offset &optional (end1 nil))
+           (read-cluster disk addr n-sectors wired-buffer)
+           (replace buffer wired-buffer :start1 base-offset :end1 end1)))
+    (do-block disk #'read-cluster lba n-sectors buffer offset)))
+
+(defmethod block-device-write ((disk sup:disk) lba n-sectors buffer &key (offset 0))
+  (flet ((write-cluster (disk addr n-sectors wired-buffer buffer base-offset &optional (end2 nil))
+           (replace wired-buffer buffer :start2 base-offset :end2 end2)
+           (write-cluster disk addr n-sectors wired-buffer)))
+    (do-block disk #'write-cluster lba n-sectors buffer offset)))
+
+(defmethod block-device-read-wired ((disk sup:disk) lba n-sectors wired-buffer)
+  (read-cluster disk lba n-sectors wired-buffer))
+
+(defmethod block-device-write-wired ((disk sup:disk) lba n-sectors wired-buffer)
+  (write-cluster disk lba n-sectors wired-buffer))
+
 (defmethod block-device-flush ((disk sup:disk))
   (sup:disk-flush disk))
 
-(defun block-device-read-sector (disk start-sector n-sectors)
-  "Read n sectors from disk"
+(defmethod block-device-read-sector ((disk sup:disk) start-sector n-sectors)
   (let ((result (make-array (* (block-device-sector-size disk) n-sectors)
                             :element-type '(unsigned-byte 8))))
     (block-device-read disk start-sector n-sectors result)
     result))
 
-(defun block-device-write-sector (disk start-sector array n-sectors)
-  "Write n sectors to disk"
+(defmethod block-device-write-sector ((disk sup:disk) start-sector array n-sectors)
   (block-device-write disk start-sector n-sectors array))
+
+;;======================================================================
+;;    Cache API implementation
+;;======================================================================
+
+(defclass page-cache (disk-partition-mixin)
+  ((%cache :initarg :cache :reader cache))
+  (:default-initargs :cache (make-hash-table :weakness :key-and-value)))
+
+(defmethod block-device-sector-size ((disk page-cache))
+  (block-device-sector-size (dp-disk disk)))
+
+(defmethod block-device-n-sectors ((disk page-cache))
+  (block-device-n-sectors (dp-disk disk)))
+
+(defmethod block-device-flush ((disk page-cache))
+  (block-device-flush (dp-disk disk)))
+
+(defmethod block-device-read-sector ((disk page-cache) start-sector n-sectors)
+  (or (gethash start-sector (cache disk))
+      (let ((wired-buffer (make-array (* (block-device-sector-size disk) n-sectors)
+                                      :element-type '(unsigned-byte 8)
+                                      :area :wired)))
+        (block-device-read-wired (dp-disk disk) start-sector n-sectors wired-buffer)
+        (setf (gethash start-sector (cache disk)) wired-buffer))))
+
+(defmethod block-device-write-sector ((disk page-cache) start-sector block n-sectors)
+  (block-device-write-wired (dp-disk disk) start-sector n-sectors (buffer block)))
```

Comment:
This doesn't seem to be used anywhere.
Using a weak key-and-value hash-table with an integer key doesn't make sense. Numbers and characters don't have sensible lifetimes for use with weak pointers. Fixnums currently live forever (but this may change in the future) so this is equivalent to using weakness value. 
Wired memory should not be kept around for longer than needed. These arrays should be allocated in normal memory if they're going to be used in a cache.
Why does it inherit from disk-partition-mixin?